### PR TITLE
12 fix wrong response when repeated requests for external source

### LINKED ISSUE #12: External source images are not responded from the repeated requests
## Test environment

In `examples/with-next-js`, I modify `main.tf` file like below:

```
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

module "next_image_optimizer" {
  source = "dealmore/next-js-image-optimization/aws"

  next_image_domains = ["assets.vercel.com","cdn11.bigcommerce.com"]
  next_image_image_sizes  = [12, 24, 48]
  next_image_device_sizes = [760, 960, 1024]
}

output "domain" {
  value = module.next_image_optimizer.cloudfront_domain_name
}
```

then `terraform apply` done successfully.

External source image: https://cdn11.bigcommerce.com/s-qfzerv205w/images/stencil/original/products/116/512/Men-Jacket-Front-Black__15466.1603283963.png

Test url: https://d1phddo1jfr778.cloudfront.net/_next/image/?url=https%3A%2F%2Fcdn11.bigcommerce.com%2Fs-qfzerv205w%2Fimages%2Fstencil%2Foriginal%2Fproducts%2F116%2F512%2FMen-Jacket-Front-Black__15466.1603283963.png&w=760&q=94

(`d1phddo1jfr778.cloudfront.net` will be destroyed in a few days)

## Expected behavior

All repeated requests could get download images.

## Current behavior

Requests with ETag header (with browser cache) works well.
But cache clear environments(without ETag) can not get images.
I use chrome browser Incognito mode for cache clear.

## Solution

I reviewed code in this repo and [nextjs image-optimizer](image-optimizer) then finally found a suspicious point.

I Think  in the `./lib/handler.ts` **handler function's return** should wait steam write. (next image-optimizer use `readStream.pipe(res)`)

@ofhouse Could I make PR?

[image-optimizer]: https://github.com/vercel/next.js/blob/master/packages/next/next-server/server/image-optimizer.ts

### PR DESCRIPTION
**Pull request by b6pzeusbc54tvhw5jgpyw8pwz2x6gs**
- fix #12 
    - Use `defer` object for awaiting steam "write".
    - `createDeffered` is used in app too, so remove `test/utils.ts`.
- add [.editorconfig](https://editorconfig.org/)

@ofhouse It's hard to set e2e:test envrionments. I ran docker container using `docker-compose.yml`, but many TCs still failed. So if you add guide for testing, I can run and add TCs.

### CODE REVIEW

**Comment by ofhouse** [comment]
Added a quick guide here for running e2e tests locally: https://github.com/dealmore/terraform-aws-next-js-image-optimization/blob/main/CONTRIBUTING.md

**Comment by ofhouse** [comment]
Whoops forgot to mention that also a local production build is required, updated the contribution guide accordingly.

**Comment by ofhouse** [comment]
@b6pzeusbc54tvhw5jgpyw8pwz2x6gs Anything here I can do to push this further?
If e2e don't work for you, I could offer to add them by myself.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
I will try to add some TCs soon.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
The problem has not been reproduced by TC, so I am checking it. A little more modification is required.

**Comment by ofhouse** [comment]
Seems like the issue is hard to track down in a e2e test, because the stream finishes before the result is returned.
Maybe a unit test where the file read is delayed would make more sense?

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
Yes, it makes sense. Another reason is that the state is initialized every `lambdaSAM.sendApiGwRequest` call in the sam local environment even the filesystem. 

```
    REPORT RequestId: 47f0c1bf-4125-471f-abba-e9e35cb24156	Init Duration: 0.34 ms	Duration: 485.10 ms	Billed Duration: 500 ms	Memory Size: 1024 MB	Max Memory Used: 1024 MB
    REPORT RequestId: fb02f7e2-c8d8-4716-9bd7-24476e97cd69	Init Duration: 0.18 ms	Duration: 385.11 ms	Billed Duration: 400 ms	Memory Size: 1024 MB	Max Memory Used: 1024 MB
```

You can see that `Init Duration` is displayed in the above continuous call log. 
I'm looking for a way to turn this initializing setting off.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
```
  const apiSpawnArgs = [
      'local', 'start-api',
      '--port', `${port}`,
      '--warm-containers', 'EAGER',
    ];
```

I have successfully reproduced by modifying the `createSAMLocal` function to above on the sammy side. Can you add `--warm-containers` option to this?

**Comment by ofhouse** [comment]
Thanks, added the warmup flag to sammy.
However when I run the test-file from your branch against the current main branch with sammy 1.4.0 the test is also passing, so it seems that it is not replicating the failing behaviour. 🙈 

But tracking down this issue further would take too much time, so I would say the PR is fine with the test.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
@ofhouse Sorry! There is a problem with that [PR](https://github.com/dealmore/sammy/pull/1). Exported function `generateSAM` should be modified to pass cli options. I will create a new PR on Sammy.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
Please check dealmore/sammy#2 before mering this PR.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
I just some "force push" to rebase latest main branch and remove unnecessary commit history. So, Be careful when checking out a branch.

**Comment by b6pzeusbc54tvhw5jgpyw8pwz2x6gs** [comment]
@ofhouse With sammy v1.5.0, I confirmed that both reproducing and resolving problem are working fine. Please check it.

File: lib/package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
   },
   "dependencies": {
     "aws-sdk": "*",
+    "change-case": "^4.1.2",
```
Comment (ofhouse) [inline lib/package.json:16]: Think this can be removed?
(See no usage in the module itself, guess this is something left from testing with sammy?)

**Review by ofhouse — CHANGES_REQUESTED** [review]
Looks good, thanks for taking the time and fixing the test! 👍 
Just one thing that was left behind, then we can merge it!

File: lib/package.json (hunk 1 above)
Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [reply lib/package.json:16]: Oh.. It looks like I added it by mistake. I will remove it tomorrow. Or you also can do it.

Comment (ofhouse) [reply lib/package.json]: No problem, resolved it 👍

**Review by ofhouse — APPROVED** [review]

**Pull request merged**